[3.0] Use integer for Amount instead of float

barryvdh opened the pull request:
See https://github.com/thephpleague/omnipay/issues/322

Breaking change; input Amount parameter is now an integer (or string representing an integer), instead of a float.

Prevent rounding/converting issues. The `getAmountInteger` is removed (as this is the default now) and a new `getAmountFloat` is added, for the old behaviour. Omnipay users would need to update their code for this.

A Money object (from https://github.com/moneyphp/money) is allows for `setMoney()`, which then sets the amount/currency on the request.

barryvdh commented:
Tests are failing because of some branch issues, but it works locally ;)

barryvdh commented:
So the Money conversion is removed, an Omnipay Amount object is added.
Currency is bound to the amount now, so no separate parameter.

frederikbosch commented:
Looks great, I'd make it `final` and change `return new static` into `return new self`.

judgej commented:
I would also give it an interface if it is going to be thrown around various classes.

barryvdh commented:
> Looks great, I'd make it final and change return new static into return new self.
Changed that. Also added some tests and tweaked the decimal conversion.

@judgej Why does it need to be an interface? Do we need to let people create their own Amount objects?

judgej commented:
Not being able to extend some features of OmniPay without hacking it has been a major frustration to me personally. So yes, if someone has an amount object that needs to do something for a specific gateway that we never imagined, then the `AmountInterface` contract would allow them to put in what functionality they want and just plug it in.

barryvdh commented:
In this case it wouldn't be for the gateway developer, but for the Omnipay user, right? Cause he decides what amount he inputs.
I'm not really sure what the exact use-case is in this situation. It would be confusing for example, to have `getFormatted()` method that returns a slightly different format then the original class.

If a gateway needs to do something specific with the amount, it can access the integer amount + currency object, and do it's own formatting.
If a user uses his own value object, it can create the Amount object based on the output of that VO.

kayladnls commented on src/Omnipay/Common/Message/AbstractRequest.php:
```diff
@@ -5,6 +5,9 @@
 
 namespace Omnipay\Common\Message;
 
+use Money\Money;
```
I think this is left by mistake from previous changes?

kayladnls commented on src/Omnipay/Common/Message/AbstractRequest.php:
```diff
@@ -301,135 +304,53 @@ public function toFloat($value)
      * Sets the payment amount.
      *
-     * @param string $value
+     * @param string|array|Amount|Money $value
```
I think this is left by mistake from previous changes?

barryvdh replied:
Yes, is fixed now.

barryvdh commented:
So one thing I'm not sure about; the default currency.

Should we still allow the default currency to be set in the gateway, and use that if one tries to set the amount without currency?
And should a `null` currency be possible?

frederikbosch commented:
@barryvdh This is a question of the Omnipay domain. You know that best of all I guess. In my opinion, if you can inject a currency and there is already a currency object, you should use that instead of the currencyCode.

```php
<?php
$gateway->purchase([
  'amount' => new Amount(100, new Currency('EUR'))
]);

$gateway->purchase([
  'amount' => new Amount(100)
]);
```

Regarding the making the currency required. The question is whether it is a problem for gateways that only recognize one currency to receive a currency. I guess not. Myself I use the Mollie gateway. I would not care to set the currency to EUR. An amount without currency does not exists in the real world. Then why should we model it differently in Omnipay?

judgej commented:
> An amount without currency does not exists in the real world.

This was my thought too.

barryvdh commented:
Okay good, then we agree on that.

And the default, currently the default is USD.

Some options:
1. Set no default, force to set the currency always explicitly
2. Allow to setCurrency on the gateway, to be able to set a default, for when just an integer is used as amount.
3. Allow the currency to be set as before (amount and value separate), transform to Amount object in the Getter, together with the Currency. This is more like it was before. `getCurrency` can be protected to prevent gateways from using it directly.

This would be something like this, with the latest commit.

```php
$request->purchase([
    'amount' => 100,
    'currency' => 'EUR',
]);

$request->getAmount()->getCurrency()->getCode() == 'EUR';

delatbabel commented:
I will point at the First Data gateways. An amount is always passed in without a currency. Each account (shop ID) processes transactions in one and only one currency. A merchant may have multiple accounts, one for each currency. In the First Data omnipay plugin code, the currency parameter is ignored. In my own payment handling code for one particular merchant (that handles multiple currencies) we have set up multiple First Data instances, each with their own shop ID and other parameters and we decide which shop ID to use based on which currency the shop wants to process. As long as I can continue doing that, this is fine by me.

judgej commented:
@delatbabel  Having a currency set for an amount won't affect any gateway that does not want to use that information. However, if that gateway wants to accept a currency code as an initialisation parameter, then it could form a validation check to make sure the wrong currency is not passed to it by accident. That is up to the gateway driver author though.

@barryvdh If the amount is an object that *has to* have a currency, then there must either be a default or it must be set explicitly each time. The idea would be that you cannot create an `Amount` object without a currency - that is the safety net. Now, if OmniPay accepts amounts that are not the `Amount` class (`AmountInterface` ;-) then some validation is needed. We don't want the validation in several places, so just throw it into a new `Amount` class and let that do the validation. The only reason to accept currency code and amount as separate fields would be to use it to create an `Amount`. Whether the currency has a default or must be explicit - I'm easy on that TBH.

barryvdh commented:
Okay, I've changed the Amount to require the currency parameter.
The currency parameter is now always required. This 4 statements will have the same result:

```php
$request->initialize(['currency' => 'USD']);
$request->purchase(['amount' => 1000]);

$request->purchase([
    'amount' => 1000,
    'currency' => 'USD',
]);

$request->purchase([
    'amount' => new Amount(1000, 'USD'),
]);

$request->purchase([
    'amount' => Amount::fromDecimal('10.00', 'USD'),
]);
```

Note that Amount/currency validation is only possible in the `getAmount()` parameter, because `currency` is not always defined on the `setAmount()`..

barryvdh commented:
@judgej Yes the Amount object can be negative/zero. It only provides check (`isNegative()` and `isZero()`), which are used in the AbstractRequest. The gateway can allow negative/zero values.

tgallice commented on src/Omnipay/Common/Message/AbstractRequest.php:
```diff
@@ -283,72 +284,65 @@ public function setCardReference($value)
      * Sets the payment amount.
      *
-     * @param string $value
+     * @param string|array|Amount $value
```
Are you sure of the `array` typehint ?

barryvdh replied:
No I removed that, leftover by mistake.
